include/cetl/pmr/memory.hpp

```diff
@@ -0,0 +1,231 @@
+    {
+        swap(std::move(rhs));
+    }
+
+    /// To support unique_ptr move semantics.
+    /// @param rhs The deleter to move from.
+    /// @return Reference to this.
+    MemoryResourceDeleter& operator=(MemoryResourceDeleter&& rhs) noexcept
+    {
+        swap(std::move(rhs));
+        return *this;
+    }
+
+    ~MemoryResourceDeleter() = default;
+
+    MemoryResourceDeleter() noexcept                               = delete;
+    MemoryResourceDeleter(const MemoryResourceDeleter&)            = delete;
+    MemoryResourceDeleter& operator=(const MemoryResourceDeleter&) = delete;
+
+    /// Functor called by smart-pointer to deallocate memory.
+    /// @param p The memory to deallocate.
+    void operator()(void* p) noexcept
+    {
+        CETL_DEBUG_ASSERT(nullptr == p || nullptr != data_.mem_resource, "mem_resource was null in deleter?");
+        if (nullptr != data_.mem_resource)
+        {
+            data_.mem_resource->deallocate(p, data_.mem_size, data_.mem_align);
+        }
+    }
+
+    /// Size of the memory this deleter will or did deallocate.
+    /// @return Size in bytes.
+    std::size_t size() const noexcept
+    {
+        return data_.mem_size;
+    }
+
+    /// Alignment of the memory this deleter will or did deallocate.
+    /// @return Alignment.
+    std::size_t alignment() const noexcept
+    {
+        return data_.mem_align;
+    }
+
+    /// The memory resource this deleter will or did use to deallocate memory.
+    /// @return The memory resource or nullptr if this deleter was moved from and is now invalid.
+    memory_resource* resource() const noexcept
+    {
+        return data_.mem_resource;
+    }
+
+private:
+    void swap(MemoryResourceDeleter&& rhs) noexcept
+    {
+        CETL_DEBUG_ASSERT(nullptr != rhs.data_.mem_resource, "Moving from resource deleter with null memory resource.");
+        data_.mem_resource = nullptr;
+        std::swap(data_, rhs.data_);
+    }
+
+    MemoryResourceData data_;
+};
+
+/// RAII helper for cetl::pf17::pmr::polymorphic_allocator and std::pmr::polymorphic_allocator. This type is designed
+/// to work with std::unique_ptr but is still cumbersome to use. Use this class with cetl::pmr::Factory for the best
+/// and safest experience.
+///
+/// @note
+/// See cetl::pmr::Factory for an example of how to use this type.
+///
+/// @tparam PolymorphicDeallocator    The type of the polymorphic allocator to use for deallocation.
+///
+template <typename PolymorphicDeallocator>
+class PolymorphicDeleter final
+{
+public:
+    using allocator  = PolymorphicDeallocator;
+    using value_type = typename PolymorphicDeallocator::value_type;
+    using unique_ptr = std::unique_ptr<value_type, cetl::pmr::PolymorphicDeleter<PolymorphicDeallocator>>;
+
+    PolymorphicDeleter() noexcept = delete;
+
+    PolymorphicDeleter(const PolymorphicDeallocator& alloc,
+                       std::size_t mem_size) noexcept(std::is_nothrow_copy_constructible<PolymorphicDeallocator>::value)
```

Comment:
"mem_size" seems misnamed since this is the number of objects not the number of bytes